pkg/spoke/controllers/finalizercontroller/manifestwork_finalize_controller.go

```diff
@@ -0,0 +1,139 @@
+	appliedManifestWorkInformer workinformer.AppliedManifestWorkInformer,
+	hubHash string,
+) factory.Controller {
+
+	controller := &ManifestWorkFinalizeController{
+		manifestWorkClient:        manifestWorkClient,
+		manifestWorkLister:        manifestWorkLister,
+		appliedManifestWorkClient: appliedManifestWorkClient,
+		appliedManifestWorkLister: appliedManifestWorkInformer.Lister(),
+		hubHash:                   hubHash,
+		rateLimiter:               workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, manifestWorkInformer.Informer()).
+		WithInformersQueueKeyFunc(helper.AppliedManifestworkQueueKeyFunc(hubHash), appliedManifestWorkInformer.Informer()).
+		WithSync(controller.sync).ToController("ManifestWorkFinalizer", recorder)
+}
+
+func (m *ManifestWorkFinalizeController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	manifestWorkName := controllerContext.QueueKey()
+	klog.V(4).Infof("Reconciling ManifestWork %q", manifestWorkName)
+
+	manifestWork, err := m.manifestWorkLister.Get(manifestWorkName)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	// If manifestwork is found, check if it is being deleted and has the finalizer.
+	if err == nil {
+		// no work to do until we're deleted
+		if manifestWork.DeletionTimestamp == nil || manifestWork.DeletionTimestamp.IsZero() {
+			return nil
+		}
+
+		// don't do work if the finalizer is not present
+		found := false
+		for i := range manifestWork.Finalizers {
+			if manifestWork.Finalizers[i] == controllers.ManifestWorkFinalizer {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil
+		}
+	}
+
+	removeFinalizer, err := m.cleanAppliedManifestWork(ctx, controllerContext, manifestWorkName)
+	if err != nil {
+		return err
+	}
+
+	// If we cannot remove manifestwork finalizer, requeue the manifestwork to check in the next loop.
+	if !removeFinalizer {
+		controllerContext.Queue().AddAfter(manifestWorkName, m.rateLimiter.When(manifestWorkName))
+		return nil
+	}
+
+	// Now we can remove manifestwork finalizer
+	// Skip if it is an orphaned appliedmanifestwork
+	if manifestWork == nil {
+		return nil
+	}
+
+	m.rateLimiter.Forget(manifestWorkName)
+	manifestWork = manifestWork.DeepCopy()
+	helper.RemoveFinalizer(manifestWork, controllers.ManifestWorkFinalizer)
+	_, err = m.manifestWorkClient.Update(ctx, manifestWork, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed to remove finalizer from ManifestWork %s/%s: %w", manifestWork.Namespace, manifestWork.Name, err)
+	}
+
+	return nil
+}
+
+// CleanAppliedManifestWork deletes appliedmanifestwork and checks if appliedmanifestwork has been cleaned.
+// It returns true only when no related appliedmanifestwork is found, otherwise tries to delete the appliedmanifestwork
```

Comment:
rather than have this method interpret the value, can it return `(appliedManifestWorkPresent bool, err error)` then the higher level can decide how to handle that?

Reply:
done